phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -318,53 +322,123 @@ public int executeTool() {
+        try (Admin admin = queryServices.getAdmin()) {
+            for (String dataTableFullName : immutableList) {
+                HashSet<String> indexes = tablesAndIndexes.get(dataTableFullName);
+                LOGGER.info("Executing " + operation + " of " + dataTableFullName
+                        + " (immutable)");
+                disableTable(admin, dataTableFullName, indexes);
+                modifyTable(admin, dataTableFullName, indexes);
+            }
+        } catch (IOException | SQLException e) {
+            LOGGER.severe("Something went wrong while disabling "
+                    + "or modifying immutable table " + e);
+        }
+        long startWaitTime = System.currentTimeMillis();
+        return startWaitTime;
+    }
+
+    private void executeToolForMutableTables(Connection conn,
+            ConnectionQueryServices queryServices,
+            Configuration conf,
+            ArrayList<String> mutableTables) {
+        LOGGER.info("Started " + operation + " for mutable tables");
+        for (String dataTableFullName : mutableTables) {
+            HashSet<String> indexes = tablesAndIndexes.get(dataTableFullName);
             try (Admin admin = queryServices.getAdmin()) {
-                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
                 LOGGER.info("Executing " + operation + " of " + dataTableFullName);
-
                 disableTable(admin, dataTableFullName, indexes);
                 modifyTable(admin, dataTableFullName, indexes);
-                if (dataTable.isImmutableRows()) {
-                    // If the table is immutable, we need to wait for clients to purge
-                    // their caches of table metadata
-                    LOGGER.info(dataTableFullName + " is an immutable table, waiting for "
-                            + (GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN + 1)
-                            + " minutes for client cache to expire");
-                    if (!(test || dryRun || indexes.isEmpty())) {
-                        try {
-                            Thread.sleep((GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN + 1)
-                                    * 60 * 1000);
-                        } catch (InterruptedException e) {
-                            LOGGER.warning("Sleep before starting index rebuild interrupted. "
-                                    + e.getMessage());
-                        }
-                    }
-                }
                 enableTable(admin, dataTableFullName, indexes);
                 LOGGER.info("Completed " + operation + " of " + dataTableFullName);
 
             } catch (IOException | SQLException e) {
-                LOGGER.severe("Something went wrong while executing " + operation
-                        + " steps " + e);
-                return -1;
+                LOGGER.severe("Something went wrong while executing "
```

Comment:
Please see the above comment for immutable tables. I think we should not ignore the failures. If this happens during downgrade/rollback, we can attempt to upgrade the table back. Maybe we can track where the failure happened and just rollback the executed operation(s). If this gets too complicated to implement, then we can re-throw the exception with a proper message (instead of ignoring the exception).

Reply:
took the first approach. Thank you for the suggestion!